x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/execution/sample/SampleIteratorTests.java

```diff
@@ -15,24 +15,38 @@
 
 import static java.util.Arrays.asList;
 import static java.util.Collections.emptyList;
-import static org.elasticsearch.xpack.eql.execution.sample.SampleIterator.matchSample;
+import static org.elasticsearch.xpack.eql.execution.sample.SampleIterator.matchSamples;
 
 public class SampleIteratorTests extends ESTestCase {
 
-    public void testMatchSample() {
+    public void testMatchSamples() {
         assertEquals(
-            asSearchHitsList(2, 1, 3),
-            matchSample(asList(asSearchHitsList(1, 1, 2), asSearchHitsList(1, 1, 1), asSearchHitsList(1, 1, 3)), 3)
+            List.of(asSearchHitsList(2, 1, 3)),
+            matchSamples(asList(asSearchHitsList(1, 1, 2), asSearchHitsList(1, 1, 1), asSearchHitsList(1, 1, 3)), 3, 1)
         );
         assertEquals(
-            asSearchHitsList(1, 4, 5),
-            matchSample(asList(asSearchHitsList(1, 2, 3), asSearchHitsList(4, 5), asSearchHitsList(4, 5, 3)), 3)
+            List.of(asSearchHitsList(1, 4, 5)),
+            matchSamples(asList(asSearchHitsList(1, 2, 3), asSearchHitsList(4, 5), asSearchHitsList(4, 5, 3)), 3, 1)
+        );
+        assertEquals(
+            List.of(asSearchHitsList(1, 2, 3)),
+            matchSamples(asList(asSearchHitsList(1), asSearchHitsList(2), asSearchHitsList(3)), 3, 1)
+        );
+        assertTrue(matchSamples(asList(asSearchHitsList(3), asSearchHitsList(3), asSearchHitsList(3)), 3, 1).isEmpty());
+        assertTrue(matchSamples(asList(asSearchHitsList(1), asSearchHitsList(1), asSearchHitsList(3)), 3, 1).isEmpty());
+        assertTrue(matchSamples(asList(asSearchHitsList(1), asSearchHitsList(3), asSearchHitsList(3)), 3, 1).isEmpty());
+        assertTrue(matchSamples(asList(asSearchHitsList(1, 1, 1), asSearchHitsList(1, 1, 1), asSearchHitsList(1, 1, 3)), 3, 1).isEmpty());
+
+        // multiple matches per key
+        assertEquals(
+            List.of(asSearchHitsList(1, 3, 4), asSearchHitsList(1, 4, 3)),
+            matchSamples(asList(asSearchHitsList(1, 2, 3), asSearchHitsList(1, 3, 4), asSearchHitsList(1, 3, 4)), 3, 2)
+        );
+
+        assertEquals(
+            List.of(asSearchHitsList(1, 3, 4), asSearchHitsList(1, 4, 3), asSearchHitsList(2, 1, 3)),
+            matchSamples(asList(asSearchHitsList(1, 2, 3), asSearchHitsList(1, 3, 4), asSearchHitsList(1, 3, 4)), 3, 3)
```

Comment:
These two tests are not enough. Incrementally increase maxSamplesPerKey to the maximum possible value and test the combinations please.